test/jdk/java/awt/font/LineBreakMeasurer/LineBreakWithTracking.java

```diff
@@ -0,0 +1,170 @@
+          + "also format the hebrew (\u05d0\u05d1\u05d2 \u05d3\u05d4\u05d5) and arabic "
+          + "(\u0627\u0628\u062a\u062c \u062e\u0644\u0627\u062e) and CJK "
+          + "(\u4e00\u4e01\u4e02\uac00\uac01\uc4fa\u67b1\u67b2\u67b3\u67b4\u67b5\u67b6\u67b7"
+          + "\u67b8\u67b9) text correctly.";
+
+  private LineBreakMeasurer lineMeasurer;
+
+  public void actionPerformed(ActionEvent e) {
+    textTracking = (float)((JRadioButtonMenuItem)e.getSource()).getClientProperty( "tracking" );
+    lineMeasurer = null;
+    invalidate();
+    repaint();
+  }
+
+  public void paintComponent(Graphics g) {
+    super.paintComponent(g);
+    setBackground(Color.white);
+
+    Graphics2D g2d = (Graphics2D)g;
+
+    if (lineMeasurer == null) {
+      Float regular = new Float(16.0);
+      Float big = new Float(24.0);
+
+      Hashtable map = new Hashtable();
+      map.put(TextAttribute.SIZE, (float)18.0);
+      map.put(TextAttribute.TRACKING, (float)textTracking);
+
+      AttributedString astr = new AttributedString(text, map);
+      astr.addAttribute(TextAttribute.SIZE, regular, 0, text.length());
+      astr.addAttribute(TextAttribute.FAMILY, fontName, 0, text.length());
+
+      int ix = text.indexOf("broken");
+      astr.addAttribute(TextAttribute.SIZE, big, ix, ix + 6);
+      ix = text.indexOf("hebrew");
+      astr.addAttribute(TextAttribute.SIZE, big, ix, ix + 6);
+      ix = text.indexOf("arabic");
+      astr.addAttribute(TextAttribute.SIZE, big, ix, ix + 6);
+      ix = text.indexOf("CJK");
+      astr.addAttribute(TextAttribute.SIZE, big, ix, ix + 3);
+
+      FontRenderContext frc = g2d.getFontRenderContext();
+      lineMeasurer = new LineBreakMeasurer(astr.getIterator(), frc);
+    }
+
+    lineMeasurer.setPosition(0);
+
+    float w = (float)getSize().width;
+    float x = 0, y = 0;
+    TextLayout layout;
+    while ((layout = lineMeasurer.nextLayout(w)) != null) {
+      x = layout.isLeftToRight() ? 0 : w - layout.getAdvance();
+      y += layout.getAscent();
+      layout.draw(g2d, x, y);
+      y += layout.getDescent() + layout.getLeading();
+    }
+  }
+}
+
+public class LineBreakWithTracking {
+
+  public void createGUI(JFrame frame) {
+
+    LineBreakPanel panel = new LineBreakPanel();
+    frame.getContentPane().add(panel, BorderLayout.CENTER);
+
+    JMenuBar menuBar = new JMenuBar();
+
+    JMenu menu = new JMenu("Tracking");
+    ButtonGroup btnGroup = new ButtonGroup();
+    String btnLabels[] = {"-0.1", "0", "0.1", "0.2", "0.3"};
+    float val = -0.1f;
+    for (String label : btnLabels) {
+      JRadioButtonMenuItem btn = new JRadioButtonMenuItem(label);
+      btn.putClientProperty( "tracking", val );
+      btn.addActionListener(panel);
+      btnGroup.add(btn);
+      menu.add(btn);
+      val += 0.1f;
+    }
+    menuBar.add(menu);
+
+    JButton btn = new JButton("Pass");
+    btn.addActionListener(new ActionListener(){
+      public void actionPerformed(ActionEvent e){
+        System.exit(0);
```

Comment:
Do not call System.exit() in jtreg tests - even manual ones or ones that run in their own VM. 
Instead dispose() the UI and then either let main() exit normally or throw RuntimeException() to indicate failure.

Also be sure that ALL exit paths clean up windows.

You may want to use PassFailJFrame.java to help you avoid writing the boilerplate and to get some of these things right

https://github.com/openjdk/jdk/blob/master/test/jdk/java/awt/regtesthelpers/PassFailJFrame.java

Reply:
fixed